engine/access/rpc/backend/scripts/scripts.go

```diff
@@ -72,67 +103,136 @@ func NewScriptsBackend(
 		state:                    state,
 		executor:                 exec,
 		maxScriptAndArgumentSize: maxScriptAndArgumentSize,
+		executionStateCache:      executionStateCache,
+		executionResultProvider:  executionResultProvider,
 	}, nil
 }
 
 // ExecuteScriptAtLatestBlock executes provided script at the latest sealed block.
+//
+// CAUTION: this layer SIMPLIFIES the ERROR HANDLING convention
+// As documented in the [access.API], which we partially implement with this function
+//   - All errors returned by this API are guaranteed to be benign. The node can continue normal operations after such errors.
+//   - Hence, we MUST check here and crash on all errors *except* for those known to be benign in the present context!
+//
+// Expected errors:
+//   - access.InvalidRequestError - the combined size (in bytes) of the script and arguments is greater than the max size
+//   - access.DataNotFoundError - no execution result info with the given block ID was found
 func (b *Scripts) ExecuteScriptAtLatestBlock(
 	ctx context.Context,
 	script []byte,
 	arguments [][]byte,
-) ([]byte, error) {
+	userCriteria optimistic_sync.Criteria,
+) ([]byte, *accessmodel.ExecutorMetadata, error) {
 	if !commonrpc.CheckScriptSize(script, arguments, b.maxScriptAndArgumentSize) {
-		return nil, status.Error(codes.InvalidArgument, commonrpc.ErrScriptTooLarge.Error())
+		return nil, nil, access.NewInvalidRequestError(commonrpc.ErrScriptTooLarge)
 	}
 
 	latestHeader, err := b.state.Sealed().Head()
 	if err != nil {
 		// the latest sealed header MUST be available
-		err := irrecoverable.NewExceptionf("failed to lookup sealed header: %w", err)
-		irrecoverable.Throw(ctx, err)
-		return nil, err
+		err = fmt.Errorf("failed to lookup latest sealed header: %w", err)
+		return nil, nil, access.RequireNoError(ctx, err)
 	}
 
-	res, _, err := b.executor.Execute(ctx, executor.NewScriptExecutionRequest(latestHeader.ID(), latestHeader.Height, script, arguments))
-	return res, err
+	executionResultInfo, err := b.executionResultProvider.ExecutionResultInfo(
+		latestHeader.ID(),
+		userCriteria,
+	)
+	if err != nil {
+		err = access.RequireErrorIs(ctx, err, storage.ErrNotFound)
+		err = fmt.Errorf("failed to get execution result info for block ID %s: %w", latestHeader.ID(), err)
+		return nil, nil, access.NewDataNotFoundError("execution result info", err)
+	}
+
+	res, metadata, _, err := b.executor.Execute(
+		ctx,
+		executor.NewScriptExecutionRequest(latestHeader.ID(), latestHeader.Height, script, arguments, executionResultInfo),
+	)
+	return res, metadata, err
```

Comment:
the intention for the new error handling framework is to have the backend endpoint implementations handle all sentinels from called methods explicitly. that way it's easy everyone working with the code to clearly see all error cases are handled correctly.

that may look something like this:
https://github.com/onflow/flow-go/blob/e22ad24de9cea9920641df149c10dc23ec207820/engine/access/rpc/backend/transactions/transactions.go#L449-L465